lib/SIL/Verifier/SILVerifier.cpp

```diff
@@ -1902,6 +1902,36 @@ class SILVerifier : public SILVerifierBase<SILVerifier> {
   void checkEndApplyInst(EndApplyInst *AI) {
     require(getAsResultOf<BeginApplyInst>(AI->getOperand())->isBeginApplyToken(),
             "operand of end_apply must be a begin_apply");
+
+    BeginApplyInst *bai = AI->getBeginApply();
+    
+    SILFunctionConventions calleeConv(bai->getSubstCalleeType(), F.getModule());
+    auto calleeResults = calleeConv.getResults();
+    auto results = AI->getResults();
+
+    require(results.size() == 1,
+            "end_apply must have a single result");
+
+    if (auto tupleTy = results[0]->getType().getAs<TupleType>()) {
+      require(tupleTy.getElementTypes().size() == calleeResults.size(),
+              "length mismatch in callee results vs. end_apply results");
+
+      for (auto typeAndIdx : llvm::enumerate(tupleTy->getElementTypes())) {
+        SILType elTy = SILType::getPrimitiveObjectType(typeAndIdx.value()->getCanonicalType());
+        requireSameType(
+          elTy,
+          calleeConv.getSILType(calleeResults[typeAndIdx.index()], F.getTypeExpansionContext()),
+          "callee result type does not match end_apply result type");
+      }
+    } else {
+      require(calleeResults.size() == 1,
+              "callee must have a single result");
+
+      requireSameType(
+        results[0]->getType(),
+        calleeConv.getSILType(calleeResults[0], F.getTypeExpansionContext()),
+        "callee result type does not match end_apply result type");
+    }
```

Comment:
You can copy the logic from `checkApplyInst` here and just use `SILFunctionConventions::getSILResultType`.

Reply:
Ouch, indeed! Somehow I completely forgot about this helper, thanks!